src/PointwiseFunctions/GeneralRelativity/ComputeGhQuantities.cpp

```diff
@@ -19,79 +19,112 @@
+    const Scalar<DataType>& lapse,
+    const tnsr::i<DataType, SpatialDim, Frame>& deriv_lapse,
     const tnsr::I<DataType, SpatialDim, Frame>& shift,
-    const tnsr::I<DataType, SpatialDim, Frame>& dt_shift,
+    const tnsr::iJ<DataType, SpatialDim, Frame>& deriv_shift,
     const tnsr::ii<DataType, SpatialDim, Frame>& spatial_metric,
-    const tnsr::ii<DataType, SpatialDim, Frame>& dt_spatial_metric,
-    const tnsr::iaa<DataType, SpatialDim, Frame>& phi) noexcept {
-  tnsr::aa<DataType, SpatialDim, Frame> pi{
-      make_with_value<DataType>(lapse, 0.)};
+    const tnsr::ijj<DataType, SpatialDim, Frame>&
+        deriv_spatial_metric) noexcept {
+  tnsr::iaa<DataType, SpatialDim, Frame> var_phi{};
+  GeneralizedHarmonic::phi<SpatialDim, Frame, DataType>(
+      make_not_null(&var_phi), lapse, deriv_lapse, shift, deriv_shift,
+      spatial_metric, deriv_spatial_metric);
+  return var_phi;
+}
+
+template <size_t SpatialDim, typename Frame, typename DataType>
+void pi(const gsl::not_null<tnsr::aa<DataType, SpatialDim, Frame>*> pi,
+        const Scalar<DataType>& lapse, const Scalar<DataType>& dt_lapse,
+        const tnsr::I<DataType, SpatialDim, Frame>& shift,
+        const tnsr::I<DataType, SpatialDim, Frame>& dt_shift,
+        const tnsr::ii<DataType, SpatialDim, Frame>& spatial_metric,
+        const tnsr::ii<DataType, SpatialDim, Frame>& dt_spatial_metric,
+        const tnsr::iaa<DataType, SpatialDim, Frame>& phi) noexcept {
+  if (UNLIKELY(get_size(get<0, 0>(*pi)) != get_size(get(lapse)))) {
+    *pi = tnsr::aa<DataType, SpatialDim, Frame>(get_size(get(lapse)));
+  }
 
-  get<0, 0>(pi) = -2. * get(lapse) * get(dt_lapse);
+  get<0, 0>(*pi) = -2. * get(lapse) * get(dt_lapse);
 
   for (size_t m = 0; m < SpatialDim; ++m) {
     for (size_t n = 0; n < SpatialDim; ++n) {
-      get<0, 0>(pi) +=
+      get<0, 0>(*pi) +=
           dt_spatial_metric.get(m, n) * shift.get(m) * shift.get(n) +
           2. * spatial_metric.get(m, n) * shift.get(m) * dt_shift.get(n);
     }
   }
 
   for (size_t i = 0; i < SpatialDim; ++i) {
+    pi->get(0, i + 1) = 0.0;
     for (size_t m = 0; m < SpatialDim; ++m) {
-      pi.get(0, i + 1) += dt_spatial_metric.get(m, i) * shift.get(m) +
-                          spatial_metric.get(m, i) * dt_shift.get(m);
+      pi->get(0, i + 1) += dt_spatial_metric.get(m, i) * shift.get(m) +
+                           spatial_metric.get(m, i) * dt_shift.get(m);
     }
     for (size_t j = i; j < SpatialDim; ++j) {
-      pi.get(i + 1, j + 1) = dt_spatial_metric.get(i, j);
+      pi->get(i + 1, j + 1) = dt_spatial_metric.get(i, j);
     }
   }
   for (size_t mu = 0; mu < SpatialDim + 1; ++mu) {
     for (size_t nu = mu; nu < SpatialDim + 1; ++nu) {
       for (size_t i = 0; i < SpatialDim; ++i) {
-        pi.get(mu, nu) -= shift.get(i) * phi.get(i, mu, nu);
+        pi->get(mu, nu) -= shift.get(i) * phi.get(i, mu, nu);
       }
-      pi.get(mu, nu) /= -get(lapse);
+      pi->get(mu, nu) /= -get(lapse);
```

Comment:
Precompute one over lapse and multiply by it here.

Reply:
this is not obvious since that requires a memory allocation. It should be measured to see which is faster depending on the number of points in the element.

Reply:
I did a few rudimentary tests in 3D for DataVectors of sizes `10` and `1000`.

For small DataVectors, I found that precomputing `one_over_lapse` gave a barely-measurable improvement over using lapse directly. By "barely-measurable" I mean that the error bars in the measurement of each case's speed _overlap_ for both alternatives. For large DataVectors, precomputing `one_over_lapse` was measurably `2-3%` slower. Therefore, I vote for *not* precomputing `one_over_lapse`.

Reply:
Since you did the measurement I think it would be good to add a C++ comment saying you did it and what you found (also, did you do the measurement on a release build? `-D CMAKE_BUILD_TYPE=Release` when running CMake)

Reply:
ping on this and you can squash

Reply:
@nilsdeppe - in Release mode, I find the following timing results for a `DataVector(5)`:
![DV5](https://user-images.githubusercontent.com/7142969/55352846-31df7100-5498-11e9-9f1a-d7e21ee97a5f.png)

and the following for a `DataVector(50)`:
![DV50](https://user-images.githubusercontent.com/7142969/55352926-69e6b400-5498-11e9-9046-598b167ccf5c.png)

Here, "outside loop" is when `one_over_lapse` is pre-computed, and "inside loop" is division by `lapse` inside the loop (which is what the code is doing now). It seems that up to `size=50`, precomputing `one_over_lapse` is somewhat slower in Release mode. I have added a comment in code to this effect + the code already has the faster alternative implemented. We should be good on this now.